app/src/main/java/io/gnosis/safe/ui/transactions/details/ButtonStateHelper.kt

```diff
@@ -1,73 +1,64 @@
 package io.gnosis.safe.ui.transactions.details
 
 class ButtonStateHelper(
+    val canSign: Boolean,
+    val hasOwnerKey: Boolean,
     val hasBeenRejected: Boolean,
-    val needsYourConfirmation: Boolean,
     val isRejection: Boolean,
-    val needsExecution: Boolean,
-    val canReject: Boolean,
-    val isOwner: Boolean,
-    val completed: Boolean
+    val awaitingConfirmations: Boolean,
+    val awaitingExecution: Boolean
 ) {
-
-    fun rejectButtonIsVisible(): Boolean =
+    fun confirmationButtonIsVisible(): Boolean =
         when {
-            !isOwner || completed -> {
-                false
-            }
-            isRejection || !canReject -> {
-                false
-            }
-            !isRejection && !needsYourConfirmation && hasBeenRejected && needsExecution -> {
-                false
-            }
-            else -> true
+            isRejection && awaitingConfirmations && canSign -> true
+            isRejection && awaitingConfirmations && hasOwnerKey && !canSign -> true
+
+            !isRejection && awaitingConfirmations && canSign -> true
+            !isRejection && awaitingConfirmations && hasOwnerKey && !canSign -> true
+
+            !isRejection && awaitingExecution && hasOwnerKey && hasBeenRejected -> false
```

Comment:
Seems a bit contrived to check for the predicates and return a hard-coded boolean value. Not sure if it would be better to return the value of the predicate directly (or its negation)  :thinking: Seems to be the style with which it was previously written, so I suppose changing it now would be out of scope.

Reply:
As a suggestion for future improvements, if you want, I think that enums or sealed classes would be a good use case to map into and then depending on the case, return `true` or `false` for visibility.

Reply:
I tried optimizing this. But it became less readable/changeable/debuggable. I don't think I understand how you would make this better. Can you give and example?